File: cli-node/src/utils/api-client.ts
import axios, { AxiosError } from "axios";
import chalk from "chalk";

export class ScopeformClientError extends Error {}
export class ScopeformAuthError extends ScopeformClientError {}
export class ScopeformForbiddenError extends ScopeformClientError {}
export class ScopeformNotFoundError extends ScopeformClientError {}
export class ScopeformConflictError extends ScopeformClientError {}
export class ScopeformAPIError extends ScopeformClientError {}

export class ScopeformClient {
  private readonly baseUrl: string;
  private readonly token?: string;

  constructor(baseUrl: string, token?: string) {
    this.baseUrl = baseUrl.replace(/\/$/, "");
    this.token = token;
  }

  private extractDetail(error: AxiosError): string {
    const payload = error.response?.data;
    if (payload && typeof payload === "object") {
      const detail = (payload as { detail?: unknown; title?: unknown }).detail;
      if (detail && typeof detail === "object") {
        const nested = detail as { detail?: unknown; title?: unknown };
        if (typeof nested.detail === "string") return nested.detail;
        if (typeof nested.title === "string") return nested.title;
      }
      if (typeof detail === "string") return detail;
      const title = (payload as { title?: unknown }).title;
      if (typeof title === "string") return title;
    }
    return "Request failed.";
  }

  private handleError(error: unknown): never {
    if (!axios.isAxiosError(error)) {
      throw new ScopeformAPIError("Unexpected API client error.");
    }

    const detail = this.extractDetail(error);
    const status = error.response?.status;
    if (status === 401) {
      console.error(chalk.bold.red(`Authentication failed: ${detail}`));
      throw new ScopeformAuthError(detail);
    }
    if (status === 403) {
      console.error(chalk.bold.red(`Permission denied: ${detail}`));
      throw new ScopeformForbiddenError(detail);
    }
    if (status === 404) {
      console.error(chalk.bold.red(`Not found: ${detail}`));
      throw new ScopeformNotFoundError(detail);
    }
    if (status === 409) {
      console.error(chalk.bold.yellow(`Conflict: ${detail}`));
      throw new ScopeformConflictError(detail);
    }

    console.error(chalk.bold.red(`API request failed (${status ?? "unknown"}): ${detail}`));
    throw new ScopeformAPIError(detail);
  }

  private async request<T>(method: "GET" | "POST", path: string, config?: Record<string, unknown>): Promise<T> {
    try {
      const response = await axios.request<T>({
        method,
        baseURL: this.baseUrl,
        url: path,
        timeout: 30_000,
        headers: {
          "Content-Type": "application/json",
          ...(this.token ? { Authorization: `Bearer ${this.token}` } : {}),
        },
        ...config,
      });
      return response.data;
    } catch (error) {
      return this.handleError(error);
    }
  }

  async exchangeAuthToken(clerkSessionToken: string): Promise<Record<string, unknown>> {
    return this.request<Record<string, unknown>>("POST", "/api/v1/auth/token", {
      data: { clerk_session_token: clerkSessionToken },
    });
  }

  async registerAgent(payload: Record<string, unknown>): Promise<Record<string, unknown>> {
    return this.request<Record<string, unknown>>("POST", "/api/v1/agents", { data: payload });
  }

  async listAgents(): Promise<Record<string, unknown>> {
    return this.request<Record<string, unknown>>("GET", "/api/v1/agents");
  }

  async getAgent(agentId: string): Promise<Record<string, unknown>> {
    return this.request<Record<string, unknown>>("GET", `/api/v1/agents/${agentId}`);
  }

  async issueToken(agentId: string, ttl: string): Promise<Record<string, unknown>> {
    return this.request<Record<string, unknown>>("POST", "/api/v1/tokens/issue", {
      data: { agent_id: agentId, ttl },
    });
  }

  async revokeToken(params: { jti?: string; agent_id?: string }): Promise<Record<string, unknown>> {
    return this.request<Record<string, unknown>>("POST", "/api/v1/tokens/revoke", { data: params });
  }

  async getLogs(params: {
    agentId?: string;
    limit?: number;
    offset?: number;
    allowed?: boolean;
    service?: string;
  }): Promise<Record<string, unknown>> {
    const query: Record<string, string | number> = {
      limit: params.limit ?? 50,
      offset: params.offset ?? 0,
    };
    if (typeof params.allowed === "boolean") {
      query.allowed = String(params.allowed);
    }
    if (params.service) {
      query.service = params.service;
    }

    return params.agentId
      ? this.request<Record<string, unknown>>("GET", `/api/v1/agents/${params.agentId}/logs`, {
          params: query,
        })
      : this.request<Record<string, unknown>>("GET", "/api/v1/logs", { params: query });
  }
}

File: cli-py/scopeform/utils/api_client.py
from __future__ import annotations

from typing import Any

import httpx
from rich.console import Console

console = Console(stderr=True)


class ScopeformClientError(Exception):
    """Base exception for CLI API client failures."""


class ScopeformAuthError(ScopeformClientError):
    """Raised for authentication failures."""


class ScopeformForbiddenError(ScopeformClientError):
    """Raised for forbidden actions."""


class ScopeformNotFoundError(ScopeformClientError):
    """Raised when a resource does not exist."""


class ScopeformConflictError(ScopeformClientError):
    """Raised when the API reports a conflict."""


class ScopeformAPIError(ScopeformClientError):
    """Raised for all other API failures."""


class ScopeformClient:
    def __init__(self, base_url: str, token: str | None = None, timeout: float = 30.0) -> None:
        self.base_url = base_url.rstrip("/")
        self.token = token
        self._client = httpx.Client(base_url=self.base_url, timeout=timeout)

    def close(self) -> None:
        self._client.close()

    def __enter__(self) -> ScopeformClient:
        return self

    def __exit__(self, exc_type, exc, tb) -> None:
        self.close()

    def _headers(self) -> dict[str, str]:
        headers = {"Content-Type": "application/json"}
        if self.token:
            headers["Authorization"] = f"Bearer {self.token}"
        return headers

    def _extract_detail(self, response: httpx.Response) -> str:
        try:
            payload = response.json()
        except ValueError:
            return "The API returned an unexpected response."

        if isinstance(payload, dict):
            detail = payload.get("detail")
            if isinstance(detail, dict):
                return str(detail.get("detail") or detail.get("title") or "Request failed.")
            if isinstance(detail, str):
                return detail
            if "title" in payload:
                return str(payload["title"])

        return "Request failed."

    def _raise_for_status(self, response: httpx.Response) -> None:
        if response.is_success:
            return

        detail = self._extract_detail(response)
        if response.status_code == 401:
            console.print(f"[bold red]Authentication failed:[/bold red] {detail}")
            raise ScopeformAuthError(detail)
        if response.status_code == 403:
            console.print(f"[bold red]Permission denied:[/bold red] {detail}")
            raise ScopeformForbiddenError(detail)
        if response.status_code == 404:
            console.print(f"[bold red]Not found:[/bold red] {detail}")
            raise ScopeformNotFoundError(detail)
        if response.status_code == 409:
            console.print(f"[bold yellow]Conflict:[/bold yellow] {detail}")
            raise ScopeformConflictError(detail)

        console.print(f"[bold red]API request failed ({response.status_code}):[/bold red] {detail}")
        raise ScopeformAPIError(detail)

    def _request(self, method: str, path: str, **kwargs: Any) -> Any:
        response = self._client.request(method, path, headers=self._headers(), **kwargs)
        self._raise_for_status(response)
        if response.status_code == 204 or not response.content:
            return None
        return response.json()

    def exchange_auth_token(self, clerk_session_token: str) -> dict[str, Any]:
        return self._request(
            "POST",
            "/api/v1/auth/token",
            json={"clerk_session_token": clerk_session_token},
        )

    def register_agent(self, payload: dict[str, Any]) -> dict[str, Any]:
        return self._request("POST", "/api/v1/agents", json=payload)

    def list_agents(self) -> dict[str, Any]:
        return self._request("GET", "/api/v1/agents")

    def get_agent(self, agent_id: str) -> dict[str, Any]:
        return self._request("GET", f"/api/v1/agents/{agent_id}")

    def issue_token(self, agent_id: str, ttl: str) -> dict[str, Any]:
        return self._request(
            "POST",
            "/api/v1/tokens/issue",
            json={"agent_id": agent_id, "ttl": ttl},
        )

    def revoke_token(self, *, jti: str | None = None, agent_id: str | None = None) -> dict[str, Any]:
        payload: dict[str, Any] = {}
        if jti is not None:
            payload["jti"] = jti
        if agent_id is not None:
            payload["agent_id"] = agent_id
        return self._request("POST", "/api/v1/tokens/revoke", json=payload)

    def get_logs(
        self,
        *,
        agent_id: str | None = None,
        limit: int = 50,
        offset: int = 0,
        allowed: bool | None = None,
        service: str | None = None,
    ) -> dict[str, Any]:
        params: dict[str, Any] = {"limit": limit, "offset": offset}
        if allowed is not None:
            params["allowed"] = str(allowed).lower()
        if service:
            params["service"] = service

        if agent_id:
            return self._request("GET", f"/api/v1/agents/{agent_id}/logs", params=params)
        return self._request("GET", "/api/v1/logs", params=params)
